In [17]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong in wrong_genres:
        df["genre"].replace(wrong, correct_genre, inplace = True)
# Itero la función sobre la columna necesaria y aplico el método replace() para modificar cada valor incorrecto inplace.

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
# Eliminar duplicados implícitos
bad_spell = ["hip", "hop", "hip-hop"]
replace_wrong_genres(bad_spell, "hiphop")
# Creo una lista con todos los posibles nombres mal escritos
# Luego aplico la función para cambiar cada nombre mal escrito por el correcto

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos
list_of_genres = df["genre"].unique()
print(sorted(list_of_genres))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`
A pesar de las grandes capacidades de Pandas, siento que los duplicados implícitos es la parte que más tiempo toma, posiblemente porque es de los pasos donde se vuelve necesaria una interpretación humana, pues en la mayoría de los casos, son errores que surgieron por humanos. También creo que es el primer momento de todo el análisis en el que es necesario tener contexto del DataFrame, poder interpretar los géneros de forma correcta, es decir, en realidad podría existir un género llamado `hip` además del `hiphop`, por ejemplo.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Esta muy bueno lo que comentas.
    
Recorda que tratamos de simular un espacio profesional donde los proyectos son dirigidos al cliente. Por lo que estos aportados de observacion o conclusion intermedia deberan usarse para explicar que cambios se realizaron en el dataset y porque, ademas de dejar en claro si se encontraron errores.
    
Esto se hace para recapitular lo realizado y darle un panorama al cliente de que problemas se podrian mejorar e informar lo realizado
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido. Te quedo excelente las observaciones.</div>

Hubo varias cosas que se revisaron antes de comenzar un análisis directo de la hipótesis. Se comenzó actualizando el nombre de las columnas: resulta que nos encontramos con unos nombres que no eran lo suficientemente explicativos de su contenido, además de algunas fallas de sintaxis como que encontramos unos espacios antes y después de los nombres, encontramos encabezados con usos de mayúsculas aleatorias.

Luego, sería importante por el rigor de la información, poder entender la razón de la falta de valores en las columnas de canciones, artistas y géneros. Esta información podría resultar útil si intentamos generar hipótesis nuevas a partir de esta primera conclusión. Las reemplazamos con la palabra `unknown` para que no obstaculice nuestro análisis, pero no lo podíamos dejar pasar por alto.

Finalmente, se revisaron algunos valores concretos de la columna `genre` en la que encontramos que habían errores en la escritura de algunos géneros.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [20]:
# Contar las canciones reproducidas en cada ciudad
df.groupby("city")["track"].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

`Comenta tus observaciones aquí`
Es evidente que se han escuchado muchas más canciones en Springfield que en Shelbyville. La diferencia es más del doble.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien.
</div>

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [21]:
# Calcular las canciones reproducidas en cada uno de los tres días
df.groupby("day_of_the_week")["track"].count()

day_of_the_week
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

`Comenta tus observaciones aquí`
No hay una diferencia evidente entre los días de la semana. Claramente la mayor cantidad de canciones se hizo el viernes, seguido del lunes, pero no es una gran diferencia.

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

¿Con respecto al miercoles?. ¿Se podria dar alguna explicacion a esto?.</div>

El miércoles resulta ser un día muy especial en nuestro análisis. Una vez analizamos las diferencias en las ciudades esto se amplía. Resulta que, en general, es el día que menos música se escuchó, pero en realidad esto sólo ocurre en la ciudad de Springfield, que cabe aclarar, es la ciudad en la que más datos encontramos, por lo que hace que se baje considerablemente la estadística. Pero en detalle, en nuestra segunda ciudad, Shelbyville, éste es justamente el día donde más registros tenemos de escucha de música. 

Entonces ahora podríamos avanzar en una investigación más intensiva que nos lleve a comprender por qué se escucha menos música los miércoles en Springfield, pero se escucha mucha más música el mismo día en Shelbyville.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [22]:
def number_tracks(day, city):
    df_day_filtered = df.loc[df["day_of_the_week"] == day]
    df_city_filtered = df_day_filtered.loc[df_day_filtered["city"] == city]
    user_count = df_city_filtered["user_id"].count()
    return user_count

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [23]:
# El número de canciones reproducidas en Springfield el lunes
print(number_tracks("Monday", "Springfield"))

15740


In [24]:
# El número de canciones reproducidas en Shelbyville el lunes
print(number_tracks("Monday", "Shelbyville"))

5614


In [25]:
# El número de canciones reproducidas en Springfield el miércoles
print(number_tracks("Wednesday", "Springfield"))

11056


In [26]:
# El número de canciones reproducidas en Shelbyville el miércoles
print(number_tracks("Wednesday", "Shelbyville"))

7003


In [27]:
# El número de canciones reproducidas en Springfield el viernes
print(number_tracks("Friday", "Springfield"))

15945


In [28]:
# El número de canciones reproducidas en Shelbyville el viernes
print(number_tracks("Friday", "Shelbyville"))

5895


In [29]:
days = [
    ["Monday", number_tracks("Monday", "Springfield"), number_tracks("Monday", "Shelbyville")],
    ["Wednesday", number_tracks("Wednesday", "Springfield"), number_tracks("Wednesday", "Shelbyville")],
    ["Friday", number_tracks("Friday", "Springfield"), number_tracks("Friday", "Shelbyville")],
]
cities = ["day", "Springfield", "Shelbyville"]
df_final = pd.DataFrame(data = days, columns = cities)
df_final

day  Springfield  Shelbyville
0     Monday        15740         5614
1  Wednesday        11056         7003
2     Friday        15945         5895

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelente idea realizar un dataframe para una mejor visualizacion.</div>

Me pareció útil poder condensar los seis resultados de una forma ordenada creando y mostrando un nuevo DataFrame en Pandas. De esa forma cualquier persona puede acceder directamente a los resultados

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`
Acepto la hipótesis completamente debido a que, como lo muestro en el DataFrame `df_final` no existe ninguna relación entre ninguna de las celdas. No se puede encontrar una relación ni un las columnas, es decir no podemos decir que se siga un patrón respecto a la ciudades, porque en Springfield se escucha más música los viernes, mientras en Shelbyville lo hacen los miércoles. Tampoco podemos construir ningún patrón en relación a las filas, porque a pesar que en Springfield haya más consumo que en Shelbyville, no hay una relación clara entre sus valores, por ejemplo el máximo de Springfield fue el viernes con 15945, mientras el máximo de Shelbyville fue el miércoles con 7003. Ahora, el mínimo de Springfield fue el miércoles y el de SHelbyville fue el lunes.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien.
</div>

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`

Acepto completamente la hipótesis. No se puede establecer una relación matemática directa entre los valores de las ciudades ni los valores de los días.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

<div class="alert alert-block alert-danger ">
<b>Comentario general</b> <a class="tocSkip"></a>

Juan, presentaste un proyecto muy bueno.

Me gusto mucho como utilizaste el codigo a traves de este trabajo y aun mas me gusto tus analisis que fueron minucios y explicativos, bajaste a tierra muchisima informacion.

El trabajo esta muy bien y queda un detalle para aprobar el mismo.

Espero tu correccion, saludos.</div>

<div class="alert alert-block alert-success ">
<b>Comentario general #2</b> <a class="tocSkip"></a>

Juan, los detalles que faltaban fueron corregidos muy bien.

Te queda un proyecto muy bien realizado y completo por lo que pasa a estar **aprobado**.

Exitos en lo que viene, saludos.</div>

[Volver a Contenidos](#back)